### ISSUE TITLE
Problem when parsing the input files

### ISSUE THREAD

**Issue by camillaugolini-iit** [EI-CoreBioinformatics/mikado#275]
Hi, I am trying to use Mikado compare with Reference file  (the gff of the Homo sapiens 98)  and Prediction file (the gff which is the output of FLAIR).
I get always the same error which I didn't get from other reference or prediction files.
I have tried to change the format of the file but the problem persists.
The error is:

2020-02-24 11:10:03,165 - main_queue - compare.py:570 - INFO - compare - MainProcess - Starting parsing the reference
2020-02-24 11:10:03,166 - main - __init__.py:125 - ERROR - main - MainProcess - list index out of range
Traceback (most recent call last):
  File "/hpcnfs/home/ieo5215/miniconda/envs/Mikado/lib/python3.6/site-packages/Mikado/__init__.py", line 110, in main
    args.func(args)
  File "/hpcnfs/home/ieo5215/miniconda/envs/Mikado/lib/python3.6/site-packages/Mikado/scales/compare.py", line 576, in compare
    ref_gff=ref_gff)
  File "/hpcnfs/home/ieo5215/miniconda/envs/Mikado/lib/python3.6/site-packages/Mikado/scales/compare.py", line 179, in prepare_reference
    gid = row.gene
  File "/hpcnfs/home/ieo5215/miniconda/envs/Mikado/lib/python3.6/site-packages/Mikado/parsers/GFF.py", line 225, in gene
    return self.parent[0]
IndexError: list index out of range

Thanks for the attention.

**Comment by lucventurini** [comment]
@camillaugolini-iit 

Thanks for reporting. I think that the problem originates from lines in the *reference* which do not have a "parent" attribute in the ninth field. Only top level features (eg genes) should be allowed not to have those.

May I ask which version of mikado was used? I think I might have fixed this issue in the development branch.

**Comment by camillaugolini-iit** [comment]
Mikado version is 1.2.4. Should I update it?

**Comment by lucventurini** [comment]
@camillaugolini-iit 

> Mikado version is 1.2.4. Should I update it?

Yes, that would be best. Would you please be able to update to either the latest [*beta* release](https://github.com/EI-CoreBioinformatics/mikado/releases/tag/2.0rc6) or directly from the `master` branch, we could verify whether the problem is still present in the current version.

I have memory of fixing related problems in the past year, so hopefully this will provide a fix.

**Comment by camillaugolini-iit** [comment]
even using the version 2.0 the error is the same.

/hpcnfs/home/ieo5215/miniconda/envs/Mikado/lib/python3.6/site-packages/sqlalchemy/sql/functions.py:67: SAWarning: The GenericFunction 'array_agg' is already registered and is going to be overriden.
  "is going to be overriden.".format(identifier)
/hpcnfs/home/ieo5215/miniconda/envs/Mikado/lib/python3.6/site-packages/Mikado/configuration/configurator.py:529: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  scoring = yaml.load(scoring_file)
2020-02-24 19:54:17,450 - main - __init__.py:124 - ERROR - main - MainProcess - Mikado crashed, cause:
2020-02-24 19:54:17,446 - main_queue - compare.py:524 - INFO - compare - MainProcess - Start
2020-02-24 19:54:17,447 - main_queue - compare.py:526 - INFO - compare - MainProcess - 
2020-02-24 19:54:17,449 - main_queue - compare.py:570 - INFO - compare - MainProcess - Starting parsing the reference
2020-02-24 19:54:17,450 - main - __init__.py:125 - ERROR - main - MainProcess - list index out of range
Traceback (most recent call last):
  File "/hpcnfs/home/ieo5215/miniconda/envs/Mikado/lib/python3.6/site-packages/Mikado/__init__.py", line 110, in main
    args.func(args)
  File "/hpcnfs/home/ieo5215/miniconda/envs/Mikado/lib/python3.6/site-packages/Mikado/scales/compare.py", line 576, in compare
    ref_gff=ref_gff)
  File "/hpcnfs/home/ieo5215/miniconda/envs/Mikado/lib/python3.6/site-packages/Mikado/scales/compare.py", line 179, in prepare_reference
    gid = row.gene
  File "/hpcnfs/home/ieo5215/miniconda/envs/Mikado/lib/python3.6/site-packages/Mikado/parsers/GFF.py", line 225, in gene
    return self.parent[0]
IndexError: list index out of range

**Comment by lucventurini** [comment]
Dear @camillaugolini-iit 
Thank you for verifying on the latest version. 
As a final favour, may I ask which human genome annotation you used? This should allow me to diagnose and fix the bug quickly. 

Kind regards

**Comment by camillaugolini-iit** [comment]
Hi, sorry for the mistake in the previous comment whic I deleted.
This time mikado is running but does not print any output. The error is the following:

2020-02-24 20:18:26,776 - main_queue - compare.py:195 - WARNING - prepare_reference - MainProcess - No gene ID found for ENST00000456328, creating a mock one.

and this is repeated for every line of the prediction file.
The genome I am using is the Human 98 (hard mask primary assembly).
The head of the prediction file is this:

1	FLAIR	transcript	14404	18664	.	-	.	ID=6c43c38f-1306-4ebc-8fb0-3a842157861b;geneID=ENSG00000227232
1	FLAIR	exon	14404	16765	.	-	.	Parent=6c43c38f-1306-4ebc-8fb0-3a842157861b
1	FLAIR	exon	16858	17742	.	-	.	Parent=6c43c38f-1306-4ebc-8fb0-3a842157861b
1	FLAIR	exon	17915	18061	.	-	.	Parent=6c43c38f-1306-4ebc-8fb0-3a842157861b
1	FLAIR	exon	18268	18664	.	-	.	Parent=6c43c38f-1306-4ebc-8fb0-3a842157861b

Thanks very much

**Comment by lucventurini** [comment]
Dear @camillaugolini-iit ,

> This time mikado is running but does not print any output. The error is the following:
`2020-02-24 20:18:26,776 - main_queue - compare.py:195 - WARNING - prepare_reference - MainProcess - No gene ID found for ENST00000456328, creating a mock one.`
and this is repeated for every line of the prediction file.

This is actually still an error in the reading of the reference file, rather than the prediction. I will try to download a version and have a go soon.

> The head of the prediction file is this:
```
1 FLAIR transcript 14404 18664 . - . ID=6c43c38f-1306-4ebc-8fb0-3a842157861b;geneID=ENSG00000227232
1 FLAIR exon 14404 16765 . - . Parent=6c43c38f-1306-4ebc-8fb0-3a842157861b
1 FLAIR exon 16858 17742 . - . Parent=6c43c38f-1306-4ebc-8fb0-3a842157861b
1 FLAIR exon 17915 18061 . - . Parent=6c43c38f-1306-4ebc-8fb0-3a842157861b
1 FLAIR exon 18268 18664 . - . Parent=6c43c38f-1306-4ebc-8fb0-3a842157861b
```

These lines by FLAIR are **invalid** GFF. Ie a transcript should not be without a gene. Giving the geneID in might well confuse Mikado; although it might still be able to manage.

I will download the GFF and let you know soon.

**Comment by lucventurini** [comment]
Dear @camillaugolini-iit 

I am unable to find in these comments the version and origin of *H. sapiens* that you are using, I am currently trying with the [GTF](https://ftp.ensembl.org/pub/release-99/gtf/homo_sapiens) and [GFF3](https://ftp.ensembl.org/pub/release-99/gff3/homo_sapiens) from Ensembl (version 38.99).

I *did* find some errors in the parsing of the GTF files, which I should have fixed in a54765b (branch `issue-275`).

If you could:
- confirm that this is indeed the *H. sapiens* version that is causing you trouble
- Test whether a54765b eliminates the warning messages when reading the reference GTF

We can verify the progress.

As for the FLAIR output, I honestly think it's a broken GFF3. I think that the best option would be to convert it to GTF using `gffread`, for a quick check.

**Comment by lucventurini** [comment]
Fixed in a37e3a3.
